packages/go_router/lib/src/builder.dart

```diff
@@ -163,34 +164,69 @@ class RouteBuilder {
 
       keyToPages.putIfAbsent(goRouteNavKey, () => <Page<dynamic>>[]).add(page);
 
-      _buildRecursive(context, matchList, startIndex + 1, pop, routerNeglect,
-          keyToPages, newParams, navigatorKey);
-    } else if (route is ShellRoute) {
+      return _buildRecursive(context, matchList, startIndex + 1, pop,
+              routerNeglect, keyToPages, newParams, navigatorKey) ??
+          state;
+    } else if (route is ShellRouteBase) {
+      if (startIndex + 1 >= matchList.matches.length) {
+        throw _RouteBuilderError('Shell routes must always have child routes');
+      }
+
       // The key for the Navigator that will display this ShellRoute's page.
       final GlobalKey<NavigatorState> parentNavigatorKey = navigatorKey;
 
-      // The key to provide to the ShellRoute's Navigator.
-      final GlobalKey<NavigatorState> shellNavigatorKey = route.navigatorKey;
-
       // Add an entry for the parent navigator if none exists.
       keyToPages.putIfAbsent(parentNavigatorKey, () => <Page<dynamic>>[]);
 
-      // Add an entry for the shell route's navigator
-      keyToPages.putIfAbsent(shellNavigatorKey, () => <Page<dynamic>>[]);
-
       // Calling _buildRecursive can result in adding pages to the
       // parentNavigatorKey entry's list. Store the current length so
       // that the page for this ShellRoute is placed at the right index.
       final int shellPageIdx = keyToPages[parentNavigatorKey]!.length;
 
-      // Build the remaining pages
-      _buildRecursive(context, matchList, startIndex + 1, pop, routerNeglect,
-          keyToPages, newParams, shellNavigatorKey);
+      // Get the current child route of this shell route from the match list.
+      final RouteBase childRoute = matchList.matches[startIndex + 1].route;
+
+      // The key to provide to the shell route's Navigator.
+      final GlobalKey<NavigatorState>? shellNavigatorKey =
+          route.navigatorKeyForChildRoute(childRoute);
+      if (shellNavigatorKey == null) {
+        throw _RouteBuilderError(
+            'Shell routes must always have a navigator key');
+      }
+
+      // Add an entry for the shell route's navigator
+      keyToPages.putIfAbsent(shellNavigatorKey, () => <Page<dynamic>>[]);
+
+      // Build the remaining pages and retrieve the state for the top of the
+      // navigation stack
+      final GoRouterState? topRouterState = _buildRecursive(
+        context,
+        matchList,
+        startIndex + 1,
+        pop,
+        routerNeglect,
+        keyToPages,
+        newParams,
+        shellNavigatorKey,
+      );
 
-      // Build the Navigator
-      final Widget child = _buildNavigator(
+      Widget child = _buildNavigator(
           pop, keyToPages[shellNavigatorKey]!, shellNavigatorKey);
```

Comment:
Currently, only the root navigator receives the `NavigatorObserver` observers list (passed on when a GoRouter object is created). Can we somehow add observers to the other navigators? I would be very grateful. Great work by the way, thanks!

Reply:
I'm not sure why the observers passed to GoRouter isn't passed down to the nested navigators. Perhaps @chunhtai can answer that? Otherwise, a field for this could of course be added to the shell route classes, if that would be preferable.

Reply:
NavigatorObserver can only be used on one Navigator at a time. it may have unexpected behavior if we pass into  sub navigators. There is an open PR to add observers parameter to shellroute https://github.com/flutter/packages/pull/2664